File: ManagedDll/Float3.cs
using System;
using System.Runtime.InteropServices;
using System.Windows.Media;

namespace Ghurund.Managed {
    [StructLayout(LayoutKind.Sequential)]
    public struct Color {
        private float x;
        private float y;
        private float z;
        private float w;

        public float R { get => x; set => x = value; }
        public float G { get => y; set => y = value; }
        public float B { get => z; set => z = value; }
        public float A { get => w; set => w = value; }

        public Color(float a, float r, float g, float b) {
            this.w = a;
            this.x = r;
            this.y = g;
            this.z = b;
        }

        public Color(string hexString) {
            System.Windows.Media.Color color2 = (System.Windows.Media.Color)ColorConverter.ConvertFromString(hexString);
            w = color2.A / 255.0f;
            x = color2.R / 255.0f;
            y = color2.G / 255.0f;
            z = color2.B / 255.0f;
        }

        public override string ToString() {
            return string.Format("#{0:X2}{1:X2}{2:X2}{3:X2}", (byte)(w * 255), (byte)(x * 255), (byte)(y * 255), (byte)(z * 255));
        }
    }

    [StructLayout(LayoutKind.Sequential)]
    public struct Float4 {
        private float x;
        private float y;
        private float z;
        private float w;

        public float X { get => x; set => x = value; }
        public float Y { get => y; set => y = value; }
        public float Z { get => z; set => z = value; }
        public float W { get => w; set => w = value; }

        public Float4(float x, float y, float z, float w) {
            this.x = x;
            this.y = y;
            this.z = z;
            this.w = w;
        }

        public override string ToString() {
            return ", " + x.ToString("0.##") + ", " + y.ToString("0.##") + ", " + z.ToString("0.##") + w.ToString("0.##");
        }
    }

    [StructLayout(LayoutKind.Sequential)]
    public struct Float3 {
        private float x;
        private float y;
        private float z;

        public float X { get => x; set => x = value; }
        public float Y { get => y; set => y = value; }
        public float Z { get => z; set => z = value; }

        public Float3(float x, float y, float z) {
            this.x = x;
            this.y = y;
            this.z = z;
        }

        public override string ToString() {
            return x.ToString("0.##") + ", " + y.ToString("0.##") + ", " + z.ToString("0.##");
        }
    }

    [StructLayout(LayoutKind.Sequential)]
    public struct Float2 {
        private float x;
        private float y;

        public float X { get => x; set => x = value; }
        public float Y { get => y; set => y = value; }

        public override string ToString() {
            return x.ToString("0.##") + ", " + y.ToString("0.##");
        }
    }

    [StructLayout(LayoutKind.Sequential)]
    public struct Int2 {
        private int x;
        private int y;

        public int X { get => x; set => x = value; }
        public int Y { get => y; set => y = value; }

        public override string ToString() {
            return x + ", " + y;
        }
    }

    [StructLayout(LayoutKind.Sequential)]
    public struct Matrix {
        public float a00;
        public float a01;
        public float a02;
        public float a03;

        public float a10;
        public float a11;
        public float a12;
        public float a13;

        public float a20;
        public float a21;
        public float a22;
        public float a23;

        public float a30;
        public float a31;
        public float a32;
        public float a33;

        public override string ToString() {
            return a00.ToString("0.##") + ", " + a01.ToString("0.##") + ", " + a02.ToString("0.##") + ", " + a03.ToString("0.##") + "\n" +
                a10.ToString("0.##") + ", " + a11.ToString("0.##") + ", " + a12.ToString("0.##") + ", " + a13.ToString("0.##") + "\n" +
                a20.ToString("0.##") + ", " + a21.ToString("0.##") + ", " + a22.ToString("0.##") + ", " + a23.ToString("0.##") + "\n" +
                a30.ToString("0.##") + ", " + a31.ToString("0.##") + ", " + a32.ToString("0.##") + ", " + a33.ToString("0.##");
        }
    }
}

File: NativeDll/src/Float3.cpp
#include "Float3.h"

XMINT2 convertInt2(const Int2 &int2) {
    return XMINT2(int2.x, int2.y);
}

Int2 convertInt2(const XMINT2 &xmInt2) {
    Int2 int2 = {xmInt2.x, xmInt2.y};
    return int2;
}

XMFLOAT2 convertFloat2(const Float2 &float2) {
    return XMFLOAT2(float2.x, float2.y);
}

Float2 convertFloat2(const XMFLOAT2 &xmFloat2) {
    Float2 float2 = {xmFloat2.x, xmFloat2.y};
    return float2;
}

XMFLOAT3 convertFloat3(const Float3 &float3) {
    return XMFLOAT3(float3.x, float3.y, float3.z);
}

Float3 convertFloat3(const XMFLOAT3 &xmFloat3) {
    Float3 float3 = {xmFloat3.x, xmFloat3.y, xmFloat3.z};
    return float3;
}

XMFLOAT4X4 convertMatrix(const Matrix &matrix) {
    return XMFLOAT4X4(matrix.a00, matrix.a01, matrix.a02, matrix.a03,
                      matrix.a10, matrix.a11, matrix.a12, matrix.a13,
                      matrix.a20, matrix.a21, matrix.a22, matrix.a23,
                      matrix.a30, matrix.a31, matrix.a32, matrix.a33);
}

Matrix convertMatrix(const XMFLOAT4X4 &xmFloat4x4) {
    Matrix matrix = {xmFloat4x4._11,xmFloat4x4._12,xmFloat4x4._13,xmFloat4x4._14,
        xmFloat4x4._21,xmFloat4x4._22,xmFloat4x4._23,xmFloat4x4._24,
        xmFloat4x4._31,xmFloat4x4._32,xmFloat4x4._33,xmFloat4x4._34,
        xmFloat4x4._41,xmFloat4x4._42,xmFloat4x4._43,xmFloat4x4._44};
    return matrix;
}